ao-tools/ao-elftohex/ao-elftohex.1
.\"
.\" Copyright © 2013 Keith Packard <keithp@keithp.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-LOAD 1 "ao-elftohex" ""
.SH NAME
ao-elftohex \- convert programs to IHX format
.SH SYNOPSIS
.B "ao-elftohex"
[\--output-\fIoutput.ihx\fP]
[\--verbose]
\fIinput.elf ...\fP
.SH DESCRIPTION
.I ao-elftohex
reads the specified .elf files and writes out a .ihx version.
.SH OPTIONS
.TP
\--output=\fIoutput.ihx\fP
This specifies the output file (default is stdout)
.TP
\--verbose
Dumps some debug information.
.TP
\--nosym
Excluded symbol table information from the resulting file. This
information is a non-standard extension supported by the Altus Metrum
tools.
.SH AUTHOR
Keith Packard


ChangeLog
commit bdabc6814b066a9b17a6557910f21b1ed03de3f0
Merge: 37dda254 c5e8826a
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Oct 28 23:10:00 2022 -0600

    Merge branch 'master' into branch-1.9

commit c5e8826aa260d4aca492d684b94adc7eea0254eb
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:07:15 2022 -0700

    Version 1.9.12
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 140a84ba8ac1e6548ab63ff586cf214d87f727a4
Author: Keith Packard <keithp@keithp.com>
Date:   Fri Oct 28 21:57:02 2022 -0700

    Auto-detect windows signing stuff
    
    Still build installers even without signing key
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7c4f9aa882256182485f7f2086221841368ddfc7
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Oct 28 22:28:40 2022 -0600

    update Releasing to include reference to jsign as a release dependency

commit 5fd95471791842234a1840477c55ba175fff5a8c
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Oct 28 22:20:03 2022 -0600

    fat: sign Windows installers with our new code signing certificate

commit ef969428f293e5155c00af514ce0782a7cdf25d5
Author: Bdale Garbee <bdale@gag.com>
Date:   Fri Oct 28 21:47:04 2022 -0600

    doc: include reference to Loki Research closures in motortest document

commit 74c73edc93110874636b4bbbe01f9f0327230b50
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:35:55 2022 -0700

    altosuilib: Remove some debug printfs
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 055803ceca56065ea2253ab3eab20a5defd00135
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:21:28 2022 -0700

    altosui: Ship motortest.pdf with packaged builds
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f57cba7dd193c5608be187d3d81781215ab38a44
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Oct 27 22:31:34 2022 -0600

    doc: additional improvements to motortest manual

commit f249b61f403b8d2a042641c5ebe0fe4a720ae436
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Oct 27 22:20:19 2022 -0600

    fix mailing list reference to use mailman3 path

commit edef072dd5cffdbd3e8346719a81808b17bdc570
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:05:10 2022 -0700

    Add EasyMotor v3 firmware to release
    
    Remove EasyMotor v2
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a4be94096281536db12378c6fee66ffb774a3eac
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:01:19 2022 -0700

    doc: Add preliminary 1.9.12 release notes
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 12b03f494462bdf8c076ea474176ec14a18026fa
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:00:47 2022 -0700

    doc: Add specs for EasyMini v3 and EasyMotor v3
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2fdf4e798bd67b0d7e96f2b8fff49915b71b089e
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 21:00:07 2022 -0700

    altosuilib: Limit EasyMotor graphs to motor pressure and accel by default
    
    Change which axes are turned on when the graph is created.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 76f0fe177a38bd16e346edf0808ee8fca343f1a2
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 20:59:20 2022 -0700

    altoslib: Map log format back to USB device id
    
    This lets us use the device_id for both telem and eeprom files.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 14a65822fa2a8bb120638633b068847c8a0197b9
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 19:39:16 2022 -0700

    altos: Allow 'c a 0' without Syntax Error
    
    Need to clear error after checking for second value in accel cal
    command.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7172638c3810f6c39e4262fa80edf5ef0dec62d8
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 19:25:02 2022 -0700

    easymotor-v3: Fix beeper config
    
    the branch had different beeper config code than mainline.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 898eeb609aa7eb078161a78ac48c727a5255a51c
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 19:19:59 2022 -0700

    ao-eeprom: Handle merge conflict from easymotor-v3 branch
    
    Had two copies of the new easymotor log format parsing bits.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 611ef4497b97bcf5239d6bc2ee13eeba9efd1503
Merge: 1e694ad4 083464b5
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 19:16:39 2022 -0700

    Merge branch 'easymotor-v3'

commit 1e694ad4f67735585f0897e2dffac755868f3150
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 17:50:36 2022 -0700

    Add EasyMini v3.0 firmware to release
    
    Followed instructions in Releasing
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit a81f5a78172663376ad9687976b9821a755d5e31
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Oct 27 17:43:08 2022 -0700

    altos: Allow setting telemetry interval to zero without crashing
    
    Setting telemetry interval to zero means disable telemetry, but the
    code was using the zero value to compute a bunch of packet delays,
    which involved dividing by the interval value causing a crash when it
    was zero. Skip those subsequent computations as they won't be used
    anyways.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 48a8ae2d38ecad5aa84b2158045c6a2fd8abe5ab
Author: Keith Packard <keithp@keithp.com>
Date:   Sun Oct 23 16:48:05 2022 -0700

    altos: Increase AES IV buffer to hold maximum AES state
    
    The compiler emits warnings if this buffer is only 16 bytes. I'm hoping
    that's because the max key size could be 256 bits instead of 128 bits?
    
    In any case, this makes the compiler happy.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3735e1eb27afb873d60164a79a9e2769dc92b3a3

ao-tools/ao-elftohex/ao-elftohex.c
/*
 * Copyright © 2013 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "ao-hex.h"
#include "ao-elf.h"
#include "ao-verbose.h"

static const struct option options[] = {
	{ .name = "verbose", .has_arg = 1, .val = 'v' },
	{ .name = "output", .has_arg = 1, .val = 'o' },
	{ .name = "nosym", .has_arg = 0, .val = 'n' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--verbose=<level>] [--output=<output.ihx>] <input.elf>\n", program);
	exit(1);
}

static int
ends_with(char *whole, char *suffix)
{
	int whole_len = strlen(whole);
	int suffix_len = strlen(suffix);

	if (suffix_len > whole_len)
		return 0;
	return strcmp(whole + whole_len - suffix_len, suffix) == 0;
}

int
main (int argc, char **argv)
{
	char			*input = NULL;
	char			*output = NULL;
	struct ao_hex_image	*full_image = NULL;
	struct ao_sym		*file_symbols = NULL;
	int			num_file_symbols;
	FILE			*file;
	int			c;
	int			i;
	int			nosym = 0;

	while ((c = getopt_long(argc, argv, "nv:o:", options, NULL)) != -1) {
		switch (c) {
		case 'o':
			output = optarg;
			break;
		case 'v':
			ao_verbose = (int) strtol(optarg, NULL, 0);
			break;
		case 'n':
			nosym = 1;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}

	if (optind >= argc)
		usage(argv[0]);

	for (i = optind; i < argc; i++) {
		struct ao_hex_image *image;

		input = argv[i];

		free(file_symbols);
		num_file_symbols = 0;
		if (ends_with (input, ".ihx"))
			image = ao_hex_load(input, &file_symbols, &num_file_symbols);
		else
			image = ao_load_elf(input, &file_symbols, &num_file_symbols);

		if (!image) {
			fprintf(stderr, "Failed to load %s\n", input);
			usage(argv[0]);
		}

		if (nosym) {
			free(file_symbols);
			file_symbols = NULL;
			num_file_symbols = 0;
		}

		if (full_image) {
			full_image = ao_hex_image_cat(full_image, image);
			if (!full_image) {
				fprintf(stderr, "Can't merge image %s\n", input);
				usage(argv[0]);
			}
		} else
			full_image = image;
	}

	if (!output)
		file = stdout;
	else {
		file = fopen(output, "w");
		if (!file) {
			perror(output);
			exit(1);
		}
	}

	if (!ao_hex_save(file, full_image, file_symbols, num_file_symbols)) {
		fprintf(stderr, "%s: failed to write hex file\n", output ? output : "<stdout>");
		if (output)
			unlink(output);
		exit(1);
	}
	exit(0);
}
